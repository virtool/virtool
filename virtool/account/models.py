from datetime import datetime

from email_validator import EmailSyntaxError, validate_email
from pydantic import ConstrainedStr, validator

from virtool_core.models.basemodel import BaseModel
from virtool_core.models.enums import AnalysisWorkflow
from virtool_core.models.group import GroupMinimal, Permissions
from virtool_core.models.user import User


def check_email(email: str | None) -> str | None:
    """Checks if the given email is valid."""
    if email is None:
        return None

    try:
        validate_email(email)
    except EmailSyntaxError:
        raise ValueError("The format of the email is invalid")

    return email


class AccountSettings(BaseModel):
    quick_analyze_workflow: AnalysisWorkflow
    show_ids: bool
    show_versions: bool
    skip_quick_analyze_dialog: bool


class ConstrainedEmail(ConstrainedStr):
    strip_whitespace = True


class Account(User):
    settings: AccountSettings
    email: ConstrainedEmail | None

    _email_validation = validator("email", allow_reuse=True)(check_email)


class APIKey(BaseModel):
    id: str
    created_at: datetime
    groups: list[GroupMinimal]
    name: str
    permissions: Permissions
