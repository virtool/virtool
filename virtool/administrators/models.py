from typing import List

from virtool_core.models.basemodel import BaseModel
from virtool_core.models.roles import AdministratorRole
from virtool_core.models.user import UserNested


class AdministratorMinimal(UserNested):
    role: AdministratorRole


class Administrator(AdministratorMinimal):
    available_roles: List[dict]


class AdministratorSearch(BaseModel):
    items: List[AdministratorMinimal]
    available_roles: List[dict]
