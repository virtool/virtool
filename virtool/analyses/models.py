import enum
from sqlalchemy import Column, Integer, String, DateTime, Enum

from virtool.postgres import Base


class AnalysisFormat(str, enum.Enum):
    """
    Enumerated type for analysis file formats

    """

    sam = "sam"
    bam = "bam"
    fasta = "fasta"
    fastq = "fastq"
    csv = "csv"
    tsv = "tsv"
    json = "json"


class AnalysisFile(Base):
    """
    SQL model to store new analysis files

    """
    __tablename__ = "analysis_files"

    id = Column(Integer, primary_key=True)
    analysis = Column(String)
    description = Column(String)
    format = Column(Enum(AnalysisFormat))
    name = Column(String)
    name_on_disk = Column(String, unique=True)
    size = Column(Integer)
    uploaded_at = Column(DateTime)

    def __repr__(self):
        return f"<AnalysisFile(id={self.id}, analysis={self.analysis}, description={self.description}, " \
               f"format={self.format}, name={self.name}, name_on_disk={self.name_on_disk}, size={self.size}, " \
               f"uploaded_at={self.uploaded_at})>"
