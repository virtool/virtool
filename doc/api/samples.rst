=======
Samples
=======

Find
----

Retrieve a list of samples, filtered by URL parameters.

::

    GET /api/samples


**Response**

.. code-block:: javascript

    [
        {
            "username": "heidi",
            "group_read": true,
            "name":"16SP017-M",
            "archived":true,
            "group_write":true,
            "_version":56,
            "nuvs":false,
            "_id":"21f6f983",
            "group":"biotech",
            "all_write":false,
            "added":"2016-07-20T15:07:52.579347",
            "imported":true,
            "pathoscope":false,
            "all_read":true
        },

        {
            "username":"heidi",
            "group_read":true,
            "name":"16SP021-M",
            "archived":true,
            "group_write":true,
            "_version":56,
            "nuvs":false,
            "_id":"92e2887d",
            "group":"biotech",
            "all_write":false,
            "added":"2016-07-20T15:09:09.299341",
            "imported":true,
            "pathoscope":false,
            "all_read":true
        }
    ]



Get
---

Retrieve a complete sample document from the server.

::

    GET /api/samples/:sample_id


Remove
------

Remove a sample document and all associated analyses.

::

    DELETE /api/samples/:sample_id


Update
------

Edit modifiable fields in a sample document.

::

    PUT /api/samples/:sample_id


**Input**

+----------+---------+----------------------------------+
| Name     | Type    | Description                      |
+==========+=========+==================================+
| name     | string  | a new name for the sample        |
+----------+---------+----------------------------------+
| host     | string  | the exact (not subtraction) host |
+----------+---------+----------------------------------+
| isolate  | string  | the originating isolate          |
+----------+---------+----------------------------------+
| archived | boolean | archive state of the sample      |
+----------+---------+----------------------------------+


Get rights
----------

Retrieve the owner group and rights for the sample.

::

    GET /api/samples/:sample_id/rights


**Response**

.. code-block:: javascript

    {
        "group": "diagnostics",
        "group_read": true,
        "group_write": true,
        "all_read": true,
        "all_write": false
    }



Modify rights
-------------

Set the rights fields and the owner group for a given sample. All fields are optional.

::

    PUT /api/samples/:sample_id/rights


**Input**

+--------------+---------+-----------------------------+
| Name         | Type    | Description                 |
+==============+=========+=============================+
| group        | string  | the owner ``group_id``      |
+--------------+---------+-----------------------------+
| group_read   | boolean | group can read sample       |
+--------------+---------+-----------------------------+
| group_write  | boolean | group can modify sample     |
+--------------+---------+-----------------------------+
| all_read     | boolean | all users can read sample   |
+--------------+---------+-----------------------------+
| all_write    | boolean | all users can modify sample |
+--------------+---------+-----------------------------+

**Response**

.. code-block:: javascript

    {
        "group": "research",
        "group_read": true,
        "group_write": false,
        "all_read": false,
        "all_write": false
    }



List analyses
-------------

Retrieve a list of all analyses associated with a given sample. The complete analysis documents are not returned.

::

    GET /api/samples/:sample_id/analyses


Analyze
-------

Start a new analysis for a given sample.

::

    POST /api/samples/:sample_id/analyses


Get analysis
------------

Retrieve a complete analysis document given a ``sample_id`` and an ``analysis_id``.

::

    GET /api/samples/:sample_id/analyses/:analysis_id


BLAST a contig
--------------

BLAST a contig that was generated as part of a NuVs analysis. This request will fail for non-NuVs analyses.

::

    POST /api/samples/:sample_id/analyses/:analysis_id/:sequence_index/blast


Remove BLAST record
-------------------

::

    DELETE /api/samples/:sample_id/analyses/:analysis_id/:sequence_index/blast


Remove analysis
---------------

Remove an analysis given a ``sample_id`` and an ``analysis_id``.

::

    DELETE /api/samples/:sample_id/analyses/:analysis_id


