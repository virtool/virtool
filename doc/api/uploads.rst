.. _uploads:

Uploads
=======

When uploads are required in Virtool, they should be targeted at ``/upload``. The available endpoints are:

+-------------------------------------------------+-------------------------+
| Situation                                       | Endpoint                |
+=================================================+=========================+
| Upload and import a Virtool viruses export file | /upload/viruses         |
+-------------------------------------------------+-------------------------+
| Upload a Illumina read file                     | /upload/reads           |
+-------------------------------------------------+-------------------------+
| Upload a .hmm file for use with NuVs            | /upload/hmm/profiles    |
+-------------------------------------------------+-------------------------+
| Upload and import a NuVs annotation file        | /upload/hmm/annotations |
+-------------------------------------------------+-------------------------+
| Upload a host FASTA file                        | /upload/host            |
+-------------------------------------------------+-------------------------+

