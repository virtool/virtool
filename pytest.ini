[pytest]
addopts = --aiohttp-loop uvloop --capture tee-sys
asyncio_mode = auto
