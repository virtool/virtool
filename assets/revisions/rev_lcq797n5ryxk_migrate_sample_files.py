"""migrate sample files

Revision ID: lcq797n5ryxk
Date: 2024-06-07 17:11:53.878606

"""

from asyncio import to_thread
from pathlib import Path

import arrow
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from virtool_core.utils import compress_file, file_stats

from virtool.migration import MigrationContext
from virtool.samples.models import SQLSampleReads
from virtool.uploads.models import SQLUpload

# Revision identifiers.
name = "migrate sample files"
created_at = arrow.get("2024-06-07 17:11:53.878606")
revision_id = "lcq797n5ryxk"

alembic_down_revision = None
virtool_down_revision = "zma2wj6b39hs"

# Change this if an Alembic revision is required to run this migration.
required_alembic_revision = None


def check_is_legacy(sample: dict[str, any]) -> bool:
    """Check if a sample has legacy read files.

    :param sample: the sample document
    :return: whether the sample is a legacy sample
    """
    return not any(file.get("raw", False) for file in sample["files"]) and all(
        file["name"] in {"reads_1.fastq", "reads_2.fastq"} for file in sample["files"]
    )


def check_is_compressed(sample: dict[str, any]) -> bool:
    """Check if a sample has compressed read files.

    :param sample: the sample document
    :return: whether the sample read files are compressed
    """
    return (
        not sample.get("is_legacy")
        or sample.get("is_compressed")
        or all(
            file["name"] in {"reads_1.fq.gz", "reads_2.fq.gz"}
            for file in sample["files"]
        )
    )


async def compress_sample_reads(
    data_path: Path,
    paths: list[Path],
    sample_files: dict[str, any],
    sample_id: int,
):
    """Compress the reads for one legacy samples.

    :param mongo: the application database object
    :param data_path: the location of the data directory
    :param files: the read files to be compressed
    :param paired: whether paired or single end reads are used
    :param sample_id: the unique identifier of the sample

    :return: a list of the compressed read files
    """
    files = []

    for i, path in enumerate(paths):
        target_filename = (
            "reads_1.fq.gz" if "reads_1.fastq" in str(path) else "reads_2.fq.gz"
        )

        target_path = data_path / "samples" / sample_id / target_filename

        await to_thread(compress_file, path, target_path, 1)

        stats = await to_thread(file_stats, target_path)

        files.append(
            {
                "name": target_filename,
                "download_url": f"/download/samples/{sample_id}/{target_filename}",
                "size": stats["size"],
                "raw": False,
                "from": sample_files[i]["from"],
            },
        )

    return files


async def upgrade(ctx: MigrationContext):
    async for sample in ctx.mongo.samples.find({"files": {"$exists": True}}):
        if "is_legacy" not in sample:
            sample["is_legacy"] = check_is_legacy(sample)

        sample_path = ctx.data_path / "samples" / sample["_id"]
        paths = [sample_path / "reads_1.fastq"]
        if sample["paired"]:
            paths.append(sample_path / "reads_2.fastq")

        if not check_is_compressed(sample):
            sample["files"] = await compress_sample_reads(
                ctx.data_path,
                paths,
                sample["files"],
                sample["_id"],
            )

        async with (
            AsyncSession(
                ctx.pg,
            ) as pg_session,
            await ctx.mongo.client.start_session() as mongo_session,
            mongo_session.start_transaction(),
        ):
            sample_id = sample["_id"]
            for file in sample["files"]:
                reads = SQLSampleReads(
                    name=file["name"],
                    name_on_disk=file["name"],
                    size=file["size"],
                    sample=sample_id,
                )

                from_ = file.get("from")

                upload = (
                    (
                        await pg_session.scalars(
                            select(SQLUpload).where(
                                SQLUpload.name_on_disk == from_["id"],
                            ),
                        )
                    )
                    .unique()
                    .one_or_none()
                )

                if not upload:
                    from_size = from_.get("size")
                    if from_size is None and (ctx.data_path / "files").exists():
                        file_path = ctx.data_path / "files" / from_.get("id")
                        from_size = (
                            file_path.stat().st_size if file_path.exists() else 0
                        )

                    upload = SQLUpload(
                        name=from_["name"],
                        name_on_disk=from_["id"],
                        size=from_size,
                        uploaded_at=from_.get("uploaded_at"),
                        removed=True,
                        reserved=True,
                    )

                upload.reads.append(reads)

                pg_session.add_all([reads, upload])

            await ctx.mongo.samples.update_one(
                {"_id": sample_id},
                {
                    "$unset": {"files": ""},
                    "$set": {"is_compressed": True, "is_legacy": sample["is_legacy"]},
                },
                session=mongo_session,
            )

            await pg_session.commit()

        for path in paths:
            await to_thread(path.unlink, missing_ok=True)
