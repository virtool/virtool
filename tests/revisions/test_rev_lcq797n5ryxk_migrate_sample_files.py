import os

import pytest
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from assets.revisions.rev_lcq797n5ryxk_migrate_sample_files import upgrade
from virtool.samples.models import SQLSampleReads
from virtool.uploads.models import SQLUpload


class TestUpgrade:
    @pytest.fixture()
    def create_files(self, ctx):
        async def func(sample):
            read_path = ctx.data_path / "samples" / sample["_id"]
            read_path.mkdir(parents=True, exist_ok=True)
            files = sample.get("files")

            if files:
                files = [files] if isinstance(files, dict) else files
                for file in files:
                    read_path.joinpath(file["name"]).write_text(file["name"])

        return func

    @pytest.fixture()
    def verify_snapshots(self, ctx, snapshot):
        async def func(sample):
            assert [sample async for sample in ctx.mongo.samples.find({})] == snapshot(
                name="mongo_after",
            )

            async with AsyncSession(ctx.pg) as session:
                assert (
                    await session.execute(select(SQLUpload))
                ).unique().scalars().all() == snapshot(name="SQLUploads after")
                assert (
                    await session.execute(select(SQLSampleReads))
                ).scalars().all() == snapshot(name="SQLSampleReads after")

            assert sorted(
                os.listdir(
                    ctx.data_path / "samples" / sample["_id"],
                ),
            ) == snapshot(
                name=f"{sample["_id"]}files after",
            )

        return func

    @pytest.fixture(autouse=True)
    async def setup_pg(self, ctx):
        async with ctx.pg.begin() as conn:
            await conn.run_sync(SQLSampleReads.metadata.create_all)
            await conn.run_sync(SQLUpload.metadata.create_all)
            await conn.commit()

    @staticmethod
    async def test_modern_sample(ctx, verify_snapshots, setup_pg):
        sample = {
            "_id": "modern_sample",
            "is_legacy": False,
            "paired": False,
        }

        read_path = ctx.data_path / "samples" / sample["_id"]
        read_path.mkdir(parents=True, exist_ok=True)

        await ctx.mongo.samples.insert_one(sample)

        await upgrade(ctx)

        await verify_snapshots(sample)

    @staticmethod
    async def test_unpaired_legacy(
        ctx,
        create_files,
        verify_snapshots,
    ):
        sample = {
            "_id": "unpaired_legacy",
            "is_legacy": True,
            "paired": False,
            "files": [
                {
                    "name": "reads_1.fastq",
                    "size": 1,
                    "raw": True,
                    "from": {
                        "id": "unpaired_legacy.fastq",
                        "name": "unpaired_legacy.fastq",
                        "size": 1,
                    },
                },
            ],
        }

        await create_files(sample)

        await ctx.mongo.samples.insert_one(sample)

        await upgrade(ctx)

        await verify_snapshots(sample)

    @staticmethod
    async def test_unpaired_legacy_pre_existing_upload(
        create_files,
        ctx,
        verify_snapshots,
        static_time,
    ):
        sample = {
            "_id": "unpaired_legacy_pre_existing_upload",
            "is_legacy": True,
            "paired": False,
            "files": [
                {
                    "name": "reads_1.fastq",
                    "size": 1,
                    "raw": True,
                    "from": {
                        "id": "unpaired_legacy_pre_existing_upload.fastq",
                        "name": "unpaired_legacy_pre_existing_upload.fastq",
                        "size": 1,
                    },
                },
            ],
        }

        async with AsyncSession(ctx.pg) as session:
            session.add(
                SQLUpload(
                    name="unpaired_legacy_pre_existing_upload",
                    name_on_disk="unpaired_legacy_pre_existing_upload.fastq",
                    size=10,
                    uploaded_at=static_time.datetime,
                    removed=False,
                    reserved=True,
                ),
            )
            await session.commit()

        await create_files(sample)

        await ctx.mongo.samples.insert_one(sample)

        await upgrade(ctx)

        await verify_snapshots(sample)

    @staticmethod
    async def test_unpaired_legacy_no_from_size(
        create_files,
        ctx,
        verify_snapshots,
    ):
        sample = {
            "_id": "unpaired_legacy_no_from_size",
            "is_legacy": True,
            "paired": False,
            "files": [
                {
                    "name": "reads_1.fastq",
                    "size": 1,
                    "raw": True,
                    "from": {
                        "id": "unpaired_legacy_no_from_size.fastq",
                        "name": "unpaired_legacy_no_from_size.fastq",
                    },
                },
            ],
        }

        await create_files(sample)

        files_path = ctx.data_path / "files"
        files_path.mkdir(parents=True, exist_ok=True)
        files_path.joinpath("unpaired_legacy_no_from_size.fastq").write_text(
            "unpaired_legacy_no_from_size",
        )

        await ctx.mongo.samples.insert_one(sample)

        await upgrade(ctx)

        await verify_snapshots(sample)

    @staticmethod
    async def test_unpaired_legacy_no_from_size_unrecoverable(
        create_files,
        ctx,
        verify_snapshots,
    ):
        sample = {
            "_id": "unpaired_legacy_no_from_size_unrecoverable",
            "is_legacy": True,
            "paired": False,
            "files": [
                {
                    "name": "reads_1.fastq",
                    "size": 1,
                    "raw": True,
                    "from": {
                        "id": "unpaired_legacy_no_from_size_unrecoverable.fastq",
                        "name": "unpaired_legacy_no_from_size_unrecoverable.fastq",
                    },
                },
            ],
        }

        files_path = ctx.data_path / "files"
        files_path.mkdir(parents=True, exist_ok=True)

        await create_files(sample)

        await ctx.mongo.samples.insert_one(sample)

        await upgrade(ctx)

        await verify_snapshots(sample)

    @staticmethod
    async def test_paired_unknown_legacy(
        create_files,
        ctx,
        verify_snapshots,
    ):
        sample = {
            "_id": "paired_unknown_legacy",
            "paired": True,
            "files": [
                {
                    "name": "reads_1.fastq",
                    "size": 1,
                    "raw": False,
                    "from": {
                        "id": "paired_unknown_legacy.fastq",
                        "name": "paired_unknown_legacy.fastq",
                        "size": 1,
                    },
                },
                {
                    "name": "reads_2.fastq",
                    "size": 1,
                    "raw": False,
                    "from": {
                        "id": "paired_unknown_legacy_2.fastq",
                        "name": "paired_unknown_legacy_2.fastq",
                        "size": 1,
                    },
                },
            ],
        }

        await create_files(sample)

        await ctx.mongo.samples.insert_one(sample)

        await upgrade(ctx)

        await verify_snapshots(sample)

    @staticmethod
    async def test_unpaired_unknown_legacy(
        create_files,
        ctx,
        verify_snapshots,
    ):
        sample = {
            "_id": "unpaired_unknown_legacy",
            "paired": False,
            "files": [
                {
                    "name": "reads_1.fastq",
                    "size": 1,
                    "raw": False,
                    "from": {
                        "id": "unpaired_unknown_legacy.fastq",
                        "name": "unpaired_unknown_legacy.fastq",
                        "size": 1,
                    },
                },
            ],
        }

        await create_files(sample)

        await ctx.mongo.samples.insert_one(sample)

        await upgrade(ctx)

        await verify_snapshots(sample)

    @staticmethod
    async def test_unpaired_legacy_compressed(
        create_files,
        ctx,
        verify_snapshots,
    ):
        sample = {
            "_id": "unpaired_legacy_compressed",
            "paired": False,
            "is_legacy": True,
            "files": [
                {
                    "name": "reads_1.fq.gz",
                    "size": 1,
                    "raw": False,
                    "from": {
                        "id": "unpaired_legacy_compressed.fastq",
                        "name": "unpaired_legacy_compressed.fastq",
                        "size": 1,
                    },
                },
            ],
        }

        await create_files(sample)

        await ctx.mongo.samples.insert_one(sample)

        await upgrade(ctx)

        await verify_snapshots(sample)

    @staticmethod
    async def test_unpaired_legacy_partial_compression(
        create_files,
        ctx,
        verify_snapshots,
    ):
        sample = {
            "_id": "unpaired_legacy_partial_compression",
            "paired": False,
            "files": [
                {
                    "name": "reads_1.fastq",
                    "size": 1,
                    "raw": False,
                    "from": {
                        "id": "unpaired_legacy_partial_compression.fastq",
                        "name": "unpaired_legacy_partial_compression.fastq",
                        "size": 1,
                    },
                },
            ],
        }

        await create_files(sample)

        (ctx.data_path / "samples" / sample["_id"] / "reads_1.fq.gz").write_text(
            sample["_id"],
        )

        await ctx.mongo.samples.insert_one(sample)

        await upgrade(ctx)

        await verify_snapshots(sample)
