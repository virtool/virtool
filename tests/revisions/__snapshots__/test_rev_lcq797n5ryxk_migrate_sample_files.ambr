# serializer version: 1
# name: TestUpgrade.test_modern_sample[SQLSampleReads after]
  list([
  ])
# ---
# name: TestUpgrade.test_modern_sample[SQLUploads after]
  list([
  ])
# ---
# name: TestUpgrade.test_modern_sample[modern_samplefiles after]
  list([
  ])
# ---
# name: TestUpgrade.test_modern_sample[mongo_after]
  list([
    dict({
      '_id': 'modern_sample',
      'is_legacy': False,
      'paired': False,
    }),
  ])
# ---
# name: TestUpgrade.test_paired_unknown_legacy[SQLSampleReads after]
  list([
    <SQLSampleReads(id=1, sample=paired_unknown_legacy, name=reads_1.fq.gz, name_on_disk=reads_1.fq.gz, size=44, upload=1, uploaded_at=None)>,
    <SQLSampleReads(id=2, sample=paired_unknown_legacy, name=reads_2.fq.gz, name_on_disk=reads_2.fq.gz, size=44, upload=2, uploaded_at=None)>,
  ])
# ---
# name: TestUpgrade.test_paired_unknown_legacy[SQLUploads after]
  list([
    <SQLUpload(id=1, created_at=None, name=paired_unknown_legacy.fastq, name_on_disk=paired_unknown_legacy.fastq, ready=False, removed=True, removed_at=None, reserved=True, size=1, space=None, type=None, user=None, uploaded_at=None)>,
    <SQLUpload(id=2, created_at=None, name=paired_unknown_legacy_2.fastq, name_on_disk=paired_unknown_legacy_2.fastq, ready=False, removed=True, removed_at=None, reserved=True, size=1, space=None, type=None, user=None, uploaded_at=None)>,
  ])
# ---
# name: TestUpgrade.test_paired_unknown_legacy[mongo_after]
  list([
    dict({
      '_id': 'paired_unknown_legacy',
      'is_compressed': True,
      'is_legacy': True,
      'paired': True,
    }),
  ])
# ---
# name: TestUpgrade.test_paired_unknown_legacy[paired_unknown_legacyfiles after]
  list([
    'reads_1.fq.gz',
    'reads_2.fq.gz',
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy[SQLSampleReads after]
  list([
    <SQLSampleReads(id=1, sample=unpaired_legacy, name=reads_1.fq.gz, name_on_disk=reads_1.fq.gz, size=44, upload=1, uploaded_at=None)>,
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy[SQLUploads after]
  list([
    <SQLUpload(id=1, created_at=None, name=unpaired_legacy.fastq, name_on_disk=unpaired_legacy.fastq, ready=False, removed=True, removed_at=None, reserved=True, size=1, space=None, type=None, user=None, uploaded_at=None)>,
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy[mongo_after]
  list([
    dict({
      '_id': 'unpaired_legacy',
      'is_compressed': True,
      'is_legacy': True,
      'paired': False,
    }),
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy[unpaired_legacyfiles after]
  list([
    'reads_1.fq.gz',
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy_compressed[SQLSampleReads after]
  list([
    <SQLSampleReads(id=1, sample=unpaired_legacy_compressed, name=reads_1.fq.gz, name_on_disk=reads_1.fq.gz, size=1, upload=1, uploaded_at=None)>,
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy_compressed[SQLUploads after]
  list([
    <SQLUpload(id=1, created_at=None, name=unpaired_legacy_compressed.fastq, name_on_disk=unpaired_legacy_compressed.fastq, ready=False, removed=True, removed_at=None, reserved=True, size=1, space=None, type=None, user=None, uploaded_at=None)>,
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy_compressed[mongo_after]
  list([
    dict({
      '_id': 'unpaired_legacy_compressed',
      'is_compressed': True,
      'is_legacy': True,
      'paired': False,
    }),
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy_compressed[unpaired_legacy_compressedfiles after]
  list([
    'reads_1.fq.gz',
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy_no_from_size[SQLSampleReads after]
  list([
    <SQLSampleReads(id=1, sample=unpaired_legacy_no_from_size, name=reads_1.fq.gz, name_on_disk=reads_1.fq.gz, size=44, upload=1, uploaded_at=None)>,
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy_no_from_size[SQLUploads after]
  list([
    <SQLUpload(id=1, created_at=None, name=unpaired_legacy_no_from_size.fastq, name_on_disk=unpaired_legacy_no_from_size.fastq, ready=False, removed=True, removed_at=None, reserved=True, size=28, space=None, type=None, user=None, uploaded_at=None)>,
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy_no_from_size[mongo_after]
  list([
    dict({
      '_id': 'unpaired_legacy_no_from_size',
      'is_compressed': True,
      'is_legacy': True,
      'paired': False,
    }),
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy_no_from_size[unpaired_legacy_no_from_sizefiles after]
  list([
    'reads_1.fq.gz',
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy_no_from_size_unrecoverable[SQLSampleReads after]
  list([
    <SQLSampleReads(id=1, sample=unpaired_legacy_no_from_size_unrecoverable, name=reads_1.fq.gz, name_on_disk=reads_1.fq.gz, size=44, upload=1, uploaded_at=None)>,
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy_no_from_size_unrecoverable[SQLUploads after]
  list([
    <SQLUpload(id=1, created_at=None, name=unpaired_legacy_no_from_size_unrecoverable.fastq, name_on_disk=unpaired_legacy_no_from_size_unrecoverable.fastq, ready=False, removed=True, removed_at=None, reserved=True, size=0, space=None, type=None, user=None, uploaded_at=None)>,
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy_no_from_size_unrecoverable[mongo_after]
  list([
    dict({
      '_id': 'unpaired_legacy_no_from_size_unrecoverable',
      'is_compressed': True,
      'is_legacy': True,
      'paired': False,
    }),
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy_no_from_size_unrecoverable[unpaired_legacy_no_from_size_unrecoverablefiles after]
  list([
    'reads_1.fq.gz',
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy_partial_compression[SQLSampleReads after]
  list([
    <SQLSampleReads(id=1, sample=unpaired_legacy_partial_compression, name=reads_1.fq.gz, name_on_disk=reads_1.fq.gz, size=44, upload=1, uploaded_at=None)>,
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy_partial_compression[SQLUploads after]
  list([
    <SQLUpload(id=1, created_at=None, name=unpaired_legacy_partial_compression.fastq, name_on_disk=unpaired_legacy_partial_compression.fastq, ready=False, removed=True, removed_at=None, reserved=True, size=1, space=None, type=None, user=None, uploaded_at=None)>,
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy_partial_compression[mongo_after]
  list([
    dict({
      '_id': 'unpaired_legacy_partial_compression',
      'is_compressed': True,
      'is_legacy': True,
      'paired': False,
    }),
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy_partial_compression[unpaired_legacy_partial_compressionfiles after]
  list([
    'reads_1.fq.gz',
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy_pre_existing_upload[SQLSampleReads after]
  list([
    <SQLSampleReads(id=1, sample=unpaired_legacy_pre_existing_upload, name=reads_1.fq.gz, name_on_disk=reads_1.fq.gz, size=44, upload=1, uploaded_at=None)>,
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy_pre_existing_upload[SQLUploads after]
  list([
    <SQLUpload(id=1, created_at=None, name=unpaired_legacy_pre_existing_upload, name_on_disk=unpaired_legacy_pre_existing_upload.fastq, ready=False, removed=False, removed_at=None, reserved=True, size=10, space=None, type=None, user=None, uploaded_at=2015-10-06 20:00:00)>,
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy_pre_existing_upload[mongo_after]
  list([
    dict({
      '_id': 'unpaired_legacy_pre_existing_upload',
      'is_compressed': True,
      'is_legacy': True,
      'paired': False,
    }),
  ])
# ---
# name: TestUpgrade.test_unpaired_legacy_pre_existing_upload[unpaired_legacy_pre_existing_uploadfiles after]
  list([
    'reads_1.fq.gz',
  ])
# ---
# name: TestUpgrade.test_unpaired_unknown_legacy[SQLSampleReads after]
  list([
    <SQLSampleReads(id=1, sample=unpaired_unknown_legacy, name=reads_1.fq.gz, name_on_disk=reads_1.fq.gz, size=44, upload=1, uploaded_at=None)>,
  ])
# ---
# name: TestUpgrade.test_unpaired_unknown_legacy[SQLUploads after]
  list([
    <SQLUpload(id=1, created_at=None, name=unpaired_unknown_legacy.fastq, name_on_disk=unpaired_unknown_legacy.fastq, ready=False, removed=True, removed_at=None, reserved=True, size=1, space=None, type=None, user=None, uploaded_at=None)>,
  ])
# ---
# name: TestUpgrade.test_unpaired_unknown_legacy[mongo_after]
  list([
    dict({
      '_id': 'unpaired_unknown_legacy',
      'is_compressed': True,
      'is_legacy': True,
      'paired': False,
    }),
  ])
# ---
# name: TestUpgrade.test_unpaired_unknown_legacy[unpaired_unknown_legacyfiles after]
  list([
    'reads_1.fq.gz',
  ])
# ---
